codersdk/agentsdk/logs.go

```diff
@@ -0,0 +1,239 @@
+	// The main context is used to close the sender goroutine and cancel
+	// any outbound requests to the API. The shudown context is used to
+	// signal the sender goroutine to flush logs and then exit.
+	ctx, cancel := context.WithCancel(context.Background())
+	shutdownCtx, shutdown := context.WithCancel(ctx)
+
+	// Synchronous sender, there can only be one outbound send at a time.
+	sendDone := make(chan struct{})
+	send := make(chan []StartupLog, 1)
+	go func() {
+		// Set flushTimeout and backlogLimit so that logs are uploaded
+		// once every 250ms or when 100 logs have been added to the
+		// backlog, whichever comes first.
+		flushTimeout := 250 * time.Millisecond
+		backlogLimit := 100
+
+		flush := time.NewTicker(flushTimeout)
+
+		var backlog []StartupLog
+		defer func() {
+			flush.Stop()
+			if len(backlog) > 0 {
+				logger.Warn(ctx, "startup logs sender exiting early, discarding logs", slog.F("discarded_logs_count", len(backlog)))
+			}
+			logger.Debug(ctx, "startup logs sender exited")
+			close(sendDone)
+		}()
+
+		done := false
+		for {
+			flushed := false
+			select {
+			case <-ctx.Done():
+				return
+			case <-shutdownCtx.Done():
+				done = true
+
+				// Check queued logs before flushing.
+				select {
+				case logs := <-send:
+					backlog = append(backlog, logs...)
+				default:
+				}
+			case <-flush.C:
+				flushed = true
+			case logs := <-send:
+				backlog = append(backlog, logs...)
+				flushed = len(backlog) >= backlogLimit
+			}
+
+			if (done || flushed) && len(backlog) > 0 {
+				flush.Stop() // Lower the chance of a double flush.
+
+				// Retry uploading logs until successful or a specific
+				// error occurs. Note that we use the main context here,
+				// meaning these requests won't be interrupted by
+				// shutdown.
+				for r := retry.New(time.Second, 5*time.Second); r.Wait(ctx); {
+					err := patchStartupLogs(ctx, PatchStartupLogs{
+						Logs: backlog,
+					})
+					if err == nil {
+						break
+					}
+
+					if errors.Is(err, context.Canceled) {
+						return
+					}
+					// This error is expected to be codersdk.Error, but it has
+					// private fields so we can't fake it in tests.
+					var statusErr interface{ StatusCode() int }
+					if errors.As(err, &statusErr) {
+						if statusErr.StatusCode() == http.StatusRequestEntityTooLarge {
+							logger.Warn(ctx, "startup logs too large, discarding logs", slog.F("discarded_logs_count", len(backlog)), slog.Error(err))
+							break
+						}
+					}
+					logger.Error(ctx, "startup logs sender failed to upload logs, retrying later", slog.F("logs_count", len(backlog)), slog.Error(err))
+				}
+				if ctx.Err() != nil {
+					return
+				}
+				backlog = nil
+
+				// Anchor flush to the last log upload.
+				flush.Reset(flushTimeout)
+			}
+			if done {
+				return
+			}
+		}
+	}()
+
+	var queue []StartupLog
+	sendLog = func(callCtx context.Context, log ...StartupLog) error {
```

Comment:
Review: Technically we don't need `callCtx` here, but right now be allow the queue/backlog to grow without bounds and if we need to limit this in the future, accepting a context here makes it much nicer for the user of the SDK, letting them abort if it takes too long.